components/TextControls/src/BaseTextAreas/MDCBaseTextArea.h

```diff
@@ -14,14 +14,125 @@
 
 #import <UIKit/UIKit.h>
 
+#import "MDCTextControlLabelBehavior.h"
+#import "MDCTextControlState.h"
+
 /**
  A UIControl subclass that leverages UITextView to provide multi-line text input
 */
-@interface MDCBaseTextArea : UIControl
+@interface MDCBaseTextArea : UIControl <UIContentSizeCategoryAdjusting>
+
+/**
+ The @c label is a label that occupies the area the text usually occupies when there is no
+ text. It is distinct from the placeholder in that it can move above the text area or disappear to
+ reveal the placeholder when editing begins.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *label;
+
+/**
+ This property determines the behavior of the textfield's label during editing.
+ @note The default is MDCTextControlLabelBehaviorFloats.
+ */
+@property(nonatomic, assign) MDCTextControlLabelBehavior labelBehavior;
+
+/**
+ The @c leadingAssistiveLabel is a label below the text on the leading edge of the view. It can be
+ used to display helper or error text.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *leadingAssistiveLabel;
 
 /**
- The UITextView contained within the text area.
+ The @c trailingAssistiveLabel is a label below the text on the trailing edge of the view. It can be
+ used to display helper or error text.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *trailingAssistiveLabel;
+
+/**
+The UITextView contained within the text area.
  */
 @property(strong, nonatomic, readonly, nonnull) UITextView *textView;
 
+/**
+The minimum number of simultaneously visible lines of text. The height of the text area will reflect
+this value if the number of lines of total text is less than or equal to this value. Fractional
+values are allowed.
+ */
+@property(nonatomic, assign) CGFloat minimumNumberOfVisibleRows;
+
+/**
+The maximum number of simultaneously visible lines of text. The height of the text area will reflect
+this value if the number of lines of total text is greater than or equal to this value. Fractional
+values are allowed.
+ */
+@property(nonatomic, assign) CGFloat maximumNumberOfVisibleRows;
+
+/**
+ Sets the floating label color for a given state. Floating label color refers to the color of the
+ label when it's in its "floating position," i.e. when it's floating.
+ @param floatingLabelColor The UIColor for the given state.
+ @param state The MDCTextControlState.
+ */
+- (void)setFloatingLabelColor:(nonnull UIColor *)floatingLabelColor
+                     forState:(MDCTextControlState)state;
```

Comment:
What is the difference between this "floating position" and editing from MDCTextControlStateEditing?

can i use [setLabelColor:blue forState:MDCTextControlStateEditing] to get the same thing? assuming we collapse this with the other normalLabel color API.

Reply:
put another way: are we ever using the floating:controlStateNormal color values? what about the normal:controlStateEditing and floating:controlStateDisabled?

Reply:
>What is the difference between this "floating position" and editing from MDCTextControlStateEditing?

MDCTextControlLabelState conveys information about the label--is it floating, is it not floating but still visible, or is it not visible at all. MDCTextControlState describes whether the text control is editing, resting (this is the normal state), or disabled.

>can i use [setLabelColor:blue forState:MDCTextControlStateEditing] to get the same thing? assuming we collapse this with the other normalLabel color API.

There was originally one stateful accessor for label color, but I had to break it up into two accessors because of requirements that the label color in certain text control states be different depending on whether the label was floating or not. This happened during the states accessibility work. I got it signed off on then, but everyone seems to have forgotten haha, based off this comment thread and a similar one in the API review last week.

>put another way: are we ever using the floating:controlStateNormal color values? what about the normal:controlStateEditing and floating:controlStateDisabled?

1. Floating label & normal state - Yes. A floating label will be visible in the normal state if there is text in the text field.
2. Normal label & editing state - No. This will never happen. You can call the setter, but you'll never see the normal label in the editing state. I don't think this is a _problem_ necessarily. I see it as a fairly innocuous quirk.
3. Floating label & disabled state - Yes. A floating label will be visible in the disabled state if there is text.

Reply:
I reread labelStateWithLabelText: and now I understand that you can have a floating non editing label.

Reply:
Sorry I dont remember, did we explore adding "floating" as a bit to the controlState?

Reply:
When you say control state you are referring to MDCTextControlState right? I'm not sure if that was considered. To me it feels like conflating two different things though. MDCTextControlLabelState, despite its name, does not really describe a state, it more describes a position--normal, floating, or none. Maybe we should rename MDCTextControlLabelState to MDCTextControlLabelPosition? That could be done as a follow up. I wouldn't want to do it in this PR because I want to keep changes to the text fields to a minimum in this PR.

Reply:
+1 MDCTextControlLabelPosition
+1 to doing this in a different PR.
I personally see "state" and "manager" poor name choice because they can mean so many things. If we can come up with more descriptive names, awesome. Naming is hard.

Reply:
I definitely see what you mean. I do stand by the name MDCTextControlState, but I'm happy to change the other one to MDCTextControlLabelPosition. I like that better.

As for the MDCTextControlGradientManager, I just called it that because nothing better came to mind. Do you have any ideas for alternatives?

Reply:
I was thinking controller but that was weak also. I didn't suggest on because I didn't come up with a good replacement.